packages/bootstrap/extra/dcos_internal_utils/utils.py

```diff
@@ -1,75 +1,139 @@
-        with open(filename) as f:
+        with filepath.open() as f:
             return json.load(f)
     except ValueError as ex:
-        raise ValueError("Invalid JSON in {0}: {1}".format(filename, ex)) from ex
+        raise ValueError("Invalid JSON in {0}: {1}".format(filepath, ex)) from ex
 
 
-def read_file_line(filename):
-    with open(filename, 'r') as f:
+def read_file_text(filepath):
+    with filepath.open() as f:
         return f.read().strip()
 
 
-# Copied from pkgpanda/util.py#L292
-def write_string(filename, data):
-    """
-    Write a string to a file.
-    Overwrite any data in that file.
+def read_file_bytes(filepath):
+    with filepath.open('rb') as f:
+        return f.read()
 
-    We use an atomic write practice of creating a temporary file and then
-    moving that temporary file to the given ``filename``. This prevents race
-    conditions such as the file being read by another process after it is
-    opened here but not yet written to.
 
-    It also prevents us from creating or truncating a file before we fail to
-    write data to it because of low disk space.
+if is_windows:
+    def write_readonly_file(filepath, data):
+        # TODO - make this atomic
+        with filepath.open('wb') as f:
+            f.write(data)
+
+    def write_private_file(filepath, data):
+        # TODO - make this atomic
+        with filepath.open('wb') as f:
+            f.write(data)
+
+    def write_public_file(filepath, data):
+        # TODO - make this atomic
+        with filepath.open('wb') as f:
+            f.write(data)
+else:
+    def write_readonly_file(filepath, data):
+        _write_file_bytes(filepath, data, 0o400)
+
+    def write_private_file(filepath, data):
+        _write_file_bytes(filepath, data, 0o600)
+
+    def write_public_file(filepath, data):
+        _write_file_bytes(filepath, data, 0o644)
+
+    def _write_file_bytes(filepath, data, mode):
+        """
+        Set the contents of file to a byte string.
+
+        The code ensures an atomic write by creating a temporary file and then
+        moving that temporary file to the given ``filename``. This prevents race
+        conditions such as the file being read by another process after it is
+        created but not yet written to.
+
+        It also prevents an invalid file being created if the `write` fails (e.g.
+        because of low disk space).
+
+        The new file is created with permissions `mode`.
+        """
+        filename = str(filepath)
+        prefix = os.path.basename(filename)
+        tmp_file_dir = os.path.dirname(os.path.realpath(filename))
+        fd, temporary_filename = tempfile.mkstemp(prefix=prefix, dir=tmp_file_dir)
+        # `mkstemp` initially creates file with permissions 0o600
+        try:
+            try:
+                os.write(fd, data)
+            finally:
+                os.close(fd)
+            os.chmod(temporary_filename, stat.S_IMODE(mode))
+            os.replace(temporary_filename, filename)
```

Comment:
Would it be a problem to move the file if it was on two different filesystems? According to docs[1]:
```
 The operation may fail if src and dst are on different filesystems. If successful, the renaming will be an atomic operation (this is a POSIX requirement).
```
[1] https://docs.python.org/3.5/library/os.html#os.replace

Reply:
The file is always on the same filesystem, since `temporary_filename` is created in the same directory as `filename`.

Reply:
Ah, sorry - I did not notice the `prefix = os.path.basename(filename)` part.